File: src/shared/native-backend/protocol.ts
import {
    NATIVE_PROTOCOL_VERSION,
    type NativeBackendHandshakeInput,
} from "./capabilities";
import type { NativeCommandName } from "./common";
import type { NativeBackendErrorPayload } from "./errors";
import type {
    NativeBackendRequestId,
    NativeProjectId,
    NativeWindowId,
    NativeWorktreeId,
} from "./ids";

export type NativeRequestMeta = {
    readonly protocolVersion: number;
    readonly sentAt?: string | null;
    readonly windowId?: NativeWindowId | null;
    readonly projectId?: NativeProjectId | null;
    readonly worktreeId?: NativeWorktreeId | null;
};

export type NativeResponseMeta = {
    readonly protocolVersion: number;
    readonly handledAt?: string | null;
};

export type NativeEventMeta = {
    readonly protocolVersion: number;
    readonly emittedAt?: string | null;
    readonly projectId?: NativeProjectId | null;
    readonly worktreeId?: NativeWorktreeId | null;
    readonly windowId?: NativeWindowId | null;
};

export type NativeRpcRequest = {
    readonly id: NativeBackendRequestId;
    readonly command: NativeCommandName;
    readonly args: Record<string, unknown>;
    readonly meta?: NativeRequestMeta;
};

export type NativeBackendResponse =
    | {
          readonly type: "response";
          readonly id: NativeBackendRequestId | null;
          readonly ok: true;
          readonly result?: unknown;
          readonly meta?: NativeResponseMeta;
      }
    | {
          readonly type: "response";
          readonly id: NativeBackendRequestId | null;
          readonly ok: false;
          readonly error: NativeBackendErrorPayload;
          readonly meta?: NativeResponseMeta;
      };

export type NativeBackendEvent = {
    readonly type: "event";
    readonly eventName: string;
    readonly payload: unknown;
    readonly meta?: NativeEventMeta;
};

export type NativeBackendOutput = NativeBackendResponse | NativeBackendEvent;

export const NATIVE_BACKEND_IPC_EVENT = "native-backend:event";

export function createNativeRequestMeta(
    overrides: Partial<NativeRequestMeta> = {},
): NativeRequestMeta {
    return {
        protocolVersion: NATIVE_PROTOCOL_VERSION,
        sentAt: new Date().toISOString(),
        windowId: null,
        projectId: null,
        worktreeId: null,
        ...overrides,
    };
}

export function createNativeHandshakeInput(
    overrides: Partial<NativeBackendHandshakeInput> = {},
): NativeBackendHandshakeInput {
    return {
        clientName: "comando-electron-main",
        clientVersion: "0.1.0",
        protocolVersion: NATIVE_PROTOCOL_VERSION,
        supportedProtocolVersions: [NATIVE_PROTOCOL_VERSION],
        ...overrides,
    };
}

export function parseNativeBackendOutput(value: unknown): NativeBackendOutput {
    if (!isRecord(value)) {
        throw new Error("Native backend output must be an object.");
    }

    if (value.type === "response") {
        return parseNativeBackendResponse(value);
    }

    if (value.type === "event") {
        if (typeof value.eventName !== "string") {
            throw new Error("Native backend eventName must be a string.");
        }

        return {
            type: "event",
            eventName: value.eventName,
            payload: value.payload,
            meta: parseOptionalEventMeta(value.meta),
        };
    }

    throw new Error("Native backend output type is not supported.");
}

function parseNativeBackendResponse(
    value: Record<string, unknown>,
): NativeBackendResponse {
    const id = parseResponseId(value.id);
    if (value.ok === true) {
        return {
            type: "response",
            id,
            ok: true,
            result: value.result,
            meta: parseOptionalResponseMeta(value.meta),
        };
    }

    if (value.ok === false) {
        if (!isRecord(value.error)) {
            throw new Error("Native backend error response is missing error.");
        }
        const error = parseNativeError(value.error);

        return {
            type: "response",
            id,
            ok: false,
            error,
            meta: parseOptionalResponseMeta(value.meta),
        };
    }

    throw new Error("Native backend response ok must be a boolean.");
}

function parseNativeError(
    value: Record<string, unknown>,
): NativeBackendErrorPayload {
    if (typeof value.code !== "string") {
        throw new Error("Native backend error code must be a string.");
    }

    if (typeof value.message !== "string") {
        throw new Error("Native backend error message must be a string.");
    }

    if (!Object.hasOwn(value, "details")) {
        throw new Error("Native backend error details is required.");
    }

    if (typeof value.retryable !== "boolean") {
        throw new Error("Native backend error retryable must be a boolean.");
    }

    return {
        code: value.code,
        message: value.message,
        details: value.details,
        retryable: value.retryable,
    };
}

function parseResponseId(value: unknown): NativeBackendRequestId | null {
    if (typeof value === "string" || typeof value === "number") {
        return value;
    }

    if (value === null) {
        return null;
    }

    throw new Error("Native backend response id must be a string, number, or null.");
}

function parseOptionalResponseMeta(value: unknown): NativeResponseMeta | undefined {
    if (value === undefined) {
        return undefined;
    }
    if (!isRecord(value) || typeof value.protocolVersion !== "number") {
        throw new Error("Native backend response meta is invalid.");
    }

    return {
        protocolVersion: value.protocolVersion,
        handledAt:
            typeof value.handledAt === "string" || value.handledAt === null
                ? value.handledAt
                : undefined,
    };
}

function parseOptionalEventMeta(value: unknown): NativeEventMeta | undefined {
    if (value === undefined) {
        return undefined;
    }
    if (!isRecord(value) || typeof value.protocolVersion !== "number") {
        throw new Error("Native backend event meta is invalid.");
    }

    return {
        protocolVersion: value.protocolVersion,
        emittedAt:
            typeof value.emittedAt === "string" || value.emittedAt === null
                ? value.emittedAt
                : undefined,
        projectId:
            typeof value.projectId === "string" || value.projectId === null
                ? value.projectId
                : undefined,
        worktreeId:
            typeof value.worktreeId === "string" || value.worktreeId === null
                ? value.worktreeId
                : undefined,
        windowId:
            typeof value.windowId === "string" || value.windowId === null
                ? value.windowId
                : undefined,
    };
}

function isRecord(value: unknown): value is Record<string, unknown> {
    return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: crates/comando-types/src/protocol.rs
use std::io::{self, Write};

use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

use crate::capabilities::PROTOCOL_VERSION;
use crate::error::{NativeError, NativeErrorCode};
use crate::ids::{ProjectId, RequestId, WindowId, WorktreeId};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeRequestMeta {
    pub protocol_version: u32,
    #[serde(default)]
    pub sent_at: Option<String>,
    #[serde(default)]
    pub window_id: Option<WindowId>,
    #[serde(default)]
    pub project_id: Option<ProjectId>,
    #[serde(default)]
    pub worktree_id: Option<WorktreeId>,
}

impl Default for NativeRequestMeta {
    fn default() -> Self {
        Self {
            protocol_version: PROTOCOL_VERSION,
            sent_at: None,
            window_id: None,
            project_id: None,
            worktree_id: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeResponseMeta {
    pub protocol_version: u32,
    #[serde(default)]
    pub handled_at: Option<String>,
}

impl Default for NativeResponseMeta {
    fn default() -> Self {
        Self {
            protocol_version: PROTOCOL_VERSION,
            handled_at: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeEventMeta {
    pub protocol_version: u32,
    #[serde(default)]
    pub emitted_at: Option<String>,
    #[serde(default)]
    pub project_id: Option<ProjectId>,
    #[serde(default)]
    pub worktree_id: Option<WorktreeId>,
    #[serde(default)]
    pub window_id: Option<WindowId>,
}

impl Default for NativeEventMeta {
    fn default() -> Self {
        Self {
            protocol_version: PROTOCOL_VERSION,
            emitted_at: None,
            project_id: None,
            worktree_id: None,
            window_id: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct NativeRpcRequest {
    pub id: RequestId,
    pub command: String,
    #[serde(default = "empty_object")]
    pub args: Value,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub meta: Option<NativeRequestMeta>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct NativeRpcResponse {
    pub id: Option<RequestId>,
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<NativeError>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub meta: Option<NativeResponseMeta>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct NativeRpcEvent {
    pub event_name: String,
    pub payload: Value,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub meta: Option<NativeEventMeta>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum NativeRpcOutput {
    #[serde(rename = "response")]
    Response(NativeRpcResponse),
    #[serde(rename = "event")]
    Event(NativeRpcEvent),
}

#[derive(Debug, Clone)]
pub struct RequestParseError {
    pub id: Option<RequestId>,
    pub error: Box<NativeError>,
}

impl RequestParseError {
    fn new(id: Option<RequestId>, error: NativeError) -> Self {
        Self {
            id,
            error: Box::new(error),
        }
    }
}

pub struct JsonlWriter<W> {
    writer: W,
}

impl<W> JsonlWriter<W>
where
    W: Write,
{
    pub fn new(writer: W) -> Self {
        Self { writer }
    }

    pub fn write_output(&mut self, output: &NativeRpcOutput) -> io::Result<()> {
        serde_json::to_writer(&mut self.writer, output).map_err(io::Error::other)?;
        self.writer.write_all(b"\n")?;
        self.writer.flush()
    }
}

pub fn parse_request_line(line: &str) -> Result<NativeRpcRequest, RequestParseError> {
    let value = serde_json::from_str::<Value>(line).map_err(|error| {
        RequestParseError::new(
            extract_id_from_malformed_json(line),
            NativeError::new(
                NativeErrorCode::InvalidJson,
                format!("Invalid JSON request: {error}"),
            ),
        )
    })?;

    parse_request_value(value)
}

pub fn parse_request_value(value: Value) -> Result<NativeRpcRequest, RequestParseError> {
    let object = value.as_object().ok_or_else(|| {
        RequestParseError::new(
            None,
            NativeError::new(
                NativeErrorCode::InvalidRequest,
                "Request must be a JSON object.",
            ),
        )
    })?;
    let id_value = object.get("id").cloned().ok_or_else(|| {
        RequestParseError::new(
            None,
            NativeError::new(NativeErrorCode::InvalidRequest, "Request id is required."),
        )
    })?;
    let id = parse_request_id(id_value).map_err(|error| RequestParseError::new(None, error))?;
    let command = object
        .get("command")
        .and_then(Value::as_str)
        .ok_or_else(|| {
            RequestParseError::new(
                Some(id.clone()),
                NativeError::new(
                    NativeErrorCode::InvalidRequest,
                    "Request command is required.",
                ),
            )
        })?;
    let args = object
        .get("args")
        .cloned()
        .unwrap_or_else(|| Value::Object(Map::new()));

    if !args.is_object() {
        return Err(RequestParseError::new(
            Some(id),
            NativeError::new(
                NativeErrorCode::InvalidRequest,
                "Request args must be an object.",
            ),
        ));
    }

    let meta = object
        .get("meta")
        .cloned()
        .map(serde_json::from_value)
        .transpose()
        .map_err(|error| {
            RequestParseError::new(
                Some(id.clone()),
                NativeError::new(
                    NativeErrorCode::InvalidRequest,
                    format!("Request meta is invalid: {error}"),
                ),
            )
        })?;

    Ok(NativeRpcRequest {
        id,
        command: command.to_string(),
        args,
        meta,
    })
}

pub fn response_ok(id: RequestId, result: Value) -> NativeRpcOutput {
    NativeRpcOutput::Response(NativeRpcResponse {
        id: Some(id),
        ok: true,
        result: Some(result),
        error: None,
        meta: None,
    })
}

pub fn error_response(id: Option<RequestId>, error: NativeError) -> NativeRpcOutput {
    NativeRpcOutput::Response(NativeRpcResponse {
        id,
        ok: false,
        result: None,
        error: Some(error),
        meta: None,
    })
}

pub fn event(event_name: impl Into<String>, payload: Value) -> NativeRpcOutput {
    NativeRpcOutput::Event(NativeRpcEvent {
        event_name: event_name.into(),
        payload,
        meta: None,
    })
}

fn empty_object() -> Value {
    Value::Object(Map::new())
}

fn parse_request_id(value: Value) -> Result<RequestId, NativeError> {
    serde_json::from_value(value).map_err(|_| {
        NativeError::new(
            NativeErrorCode::InvalidRequest,
            "Request id must be a string or number.",
        )
    })
}

fn extract_id_from_malformed_json(line: &str) -> Option<RequestId> {
    let id_key_start = line.find("\"id\"")?;
    let colon_offset = line[id_key_start + 4..].find(':')?;
    let value_start = id_key_start + 4 + colon_offset + 1;
    let value_text = line[value_start..].trim_start();

    if value_text.starts_with('"') {
        return parse_string_id(value_text);
    }

    let token: String = value_text
        .chars()
        .take_while(|character| !matches!(character, ',' | '}' | ']') && !character.is_whitespace())
        .collect();
    if token.is_empty() {
        return None;
    }

    serde_json::from_str::<RequestId>(&token).ok()
}

fn parse_string_id(value_text: &str) -> Option<RequestId> {
    let mut escaped = false;

    for (index, character) in value_text.char_indices().skip(1) {
        if escaped {
            escaped = false;
            continue;
        }

        match character {
            '\\' => escaped = true,
            '"' => return serde_json::from_str(&value_text[..=index]).ok(),
            _ => {}
        }
    }

    None
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn req_id(value: i64) -> RequestId {
        RequestId::Number(value)
    }

    #[test]
    fn parses_request_with_args() {
        let request = parse_request_line(r#"{"id":1,"command":"backend_ping","args":{}}"#)
            .expect("request should parse");

        assert_eq!(request.id, req_id(1));
        assert_eq!(request.command, "backend_ping");
        assert_eq!(request.args, json!({}));
    }

    #[test]
    fn parses_request_without_args() {
        let request = parse_request_line(r#"{"id":"abc","command":"backend_ping"}"#)
            .expect("request should parse");

        assert_eq!(request.id, RequestId::String("abc".to_string()));
        assert_eq!(request.args, json!({}));
    }

    #[test]
    fn reports_invalid_json_with_id_when_possible() {
        let error = parse_request_line(r#"{"id":42,"command":"backend_ping""#)
            .expect_err("request should fail");

        assert_eq!(error.id, Some(req_id(42)));
        assert_eq!(error.error.code, NativeErrorCode::InvalidJson);
    }

    #[test]
    fn rejects_non_object_args() {
        let error = parse_request_line(r#"{"id":1,"command":"backend_ping","args":[]}"#)
            .expect_err("request should fail");

        assert_eq!(error.id, Some(req_id(1)));
        assert_eq!(error.error.code, NativeErrorCode::InvalidRequest);
    }

    #[test]
    fn serializes_success_response() {
        let output = response_ok(req_id(1), json!({"pong": true}));
        let serialized = serde_json::to_value(output).expect("response should serialize");

        assert_eq!(
            serialized,
            json!({"type":"response","id":1,"ok":true,"result":{"pong":true}})
        );
    }

    #[test]
    fn serializes_error_response() {
        let output = error_response(
            Some(req_id(1)),
            NativeError::new(NativeErrorCode::UnknownCommand, "Unknown command: missing"),
        );
        let serialized = serde_json::to_value(output).expect("response should serialize");

        assert_eq!(
            serialized,
            json!({
                "type":"response",
                "id":1,
                "ok":false,
                "error":{
                    "code":"unknown_command",
                    "message":"Unknown command: missing",
                    "details":null,
                    "retryable":false
                }
            })
        );
    }

    #[test]
    fn serializes_event() {
        let output = event("backend://test-event", json!({"message":"hello"}));
        let serialized = serde_json::to_value(output).expect("event should serialize");

        assert_eq!(
            serialized,
            json!({
                "type":"event",
                "eventName":"backend://test-event",
                "payload":{"message":"hello"}
            })
        );
    }
}
